# Homework 7
## Exercise - Conceptuals
### 3a) Explain how k-fold cross-validation is implemented.

> K-fold CV involves randomly dividing the set of observations into k groups, or folds, of approximately equal size. The first fold is treated as a validation set (test set), and the method is fit on the remaining k − 1 folds. We then calculate the mean squared error on the observations in the held-out fold.
This procedure is repeated k times; each time, a different group of observations is treated as a validation set. This process results in k estimates of the test error, MSE1, MSE2,...,MSEk.

### 3b) General advantages & disadvantages

### The validation set approach - advantages:

> Although we dont know which of the curves, derived from different test MSE estimate for each of the regression models consider results in the smallest validation set MSE - we do know if for example a linear fit is adequate for our data. That's helpful!

### The validation set approach - disadvantages:

> The variability of the validation estimates of the test error rate can be very high, depending on which observations are included in the training set and which observations are included in the validation set.

> Only a subset of the observations — those included in the training set rather than in the validation set — are used to fit the model. This suggests that the validation set error rate may tend to overestimate the test error rate for the model fit on the entire data set.

> We dont know which of the curves, derived from different test MSE estimate for each of the regression models consider results in the smallest validation set MSE - we only know if for example a linear fit is adequate for our data. 


### The LOOCV approach - advantages:

> It provides an approximately unbiased estimate for the test error. It has far less bias than the validation set approach. 

> The LOOCV approach tends not to overestimate the test error rate as much as the validation set approach does.

> when applied repeatedly it will always yield the same results: there is no randomness in the training/vali- dation set splits.

### The LOOCV approach - disadvantages:

> LOOCV has the potential to be expensive (in time) to implement, since the model
has to be fit n times. 

# and now to the actual question...
### What are the advantages and disadvantages of k-fold cross- validation relative to: 
### validation set approach:

> (+) There is typically some variability in the CV estimates as a result of the variability in how the observations are divided into ten folds. But this variability is typically much lower than the variability in the test error estimates that results from the validation set approach. 

> (+) K-fold CV doesn't overestimate the test error rate, compared to the validation set approach. 

> (-) Can be more timeconsuming, when choosing a high k.


### What are the advantages and disadvantages of k-fold cross- validation relative to: 
### LOOCV:

> (+) LOOCV requires fitting the statistical learning method n times. This has the potential to be computationally expensive. But cross-validation is a very general approach that can be applied to almost any statistical learning method. CV requires fitting the learning procedure only ten times (if k = 10), which may be much more feasible.


> (+) / (-) The main advantage of k-fold is the bias-variance trade-off: k-fold CV often gives more accurate estimates of the test error rate than does LOOCV. But LOOCV is less biased since there are as many training sets as observations. However, LOOCV has higher variance than k-fold CV does with k < n.
Compared to the validation set approach, LOOCV provides an approximately unbiased estimate for the test error/ less bias than the validation set approach. 

## Exercise - Applied
### 7a) Fit a logistic regressionmodel that predicts Direction using Lag1 and Lag2.

In [7]:
install.packages("ISLR")
library(ISLR)
head(Weekly)
set.seed(1)
attach(Weekly)


The downloaded binary packages are in
	/var/folders/sc/hyjdb_fs3q9bfhvw0h8h4t9m0000gn/T//RtmpjqfK6N/downloaded_packages


Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down

The following objects are masked from Weekly (pos = 3):

    Direction, Lag1, Lag2, Lag3, Lag4, Lag5, Today, Volume, Year




In [10]:
glm1 = glm(Direction~Lag1+Lag2, data=Weekly, family=binomial)
summary(glm.fit)
coef(glm1)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


(Intercept)        Lag1        Lag2 
 0.22122405 -0.03872222  0.06024830

### 7b) Fit a logistic regressionmodel that predicts Direction using Lag1 and Lag2, using all but the first observation

In [11]:
glm2 <- glm(Direction ~ Lag1 + Lag2, data = Weekly[-1, ], family = binomial)
summary(glm2)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly[-1, 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


### 7c) Predict the direction of the first observation. Was it correctly classified?

In [18]:
predict.glm(glm2, Weekly[1,],Direction="Up"|Lag1, Lag2, type="response") > 0.5
predict.glm(glm2, Weekly[1,], type="response") > 0.5

Warning message in if (!se.fit) {:
“Bedingung hat Länge > 1 und nur das erste Element wird benutzt”
Warning message in if (se.fit) list(fit = predictor, se.fit = se, df = df, residual.scale = sqrt(res.var)) else predictor:
“Bedingung hat Länge > 1 und nur das erste Element wird benutzt”


fit         se.fit residual.scale 
          TRUE          FALSE           TRUE

1 
TRUE

> No, it's not correctly classified: The prediction was "Up" but it actually goes "Down".

### 7d) Write a for loop from i =1 to i = n, and perform following steps:
- Fit a logistic regression model using all but the ith obser- vation to predict Direction using Lag1 and Lag2.
- Compute the posterior probability of the market moving up for the ith observation.
- Use the posterior probability for the ith observation in order to predict whether or not the market moves up.
- Determine whether or not an error was made in predicting the direction for the ith observation.

In [28]:
count = rep(0, dim(Weekly)[1])
for (i in 1:(dim(Weekly)[1])) {
    glm3 = glm(Direction~Lag1+Lag2, data=Weekly[-i,], family=binomial)
    check_up = predict.glm(glm3, Weekly[i,], type ="response") > 0.5
    true_up = Weekly[i,]$Direction == "Up"
    if (check_up != true_up) 
    count[i] = 1
    else
        0
    }
sum(count)

[1] 490

### 7e) Take the average of the n numbers obtained in (d)iv in order to obtain the LOOCV estimate for the test error. Comment on the results.


In [29]:
round(mean(count),3)

[1] 0.45

> The LOOCV approach gives an estimated test error rate of 45%.


### 9a) Provide an estimate for the population mean of medv.

In [30]:
library(boot)
library(MASS)
attach(Boston)
str(Boston)
set.seed(1)

'data.frame':	506 obs. of  14 variables:
 $ crim   : num  0.00632 0.02731 0.02729 0.03237 0.06905 ...
 $ zn     : num  18 0 0 0 0 0 12.5 12.5 12.5 12.5 ...
 $ indus  : num  2.31 7.07 7.07 2.18 2.18 2.18 7.87 7.87 7.87 7.87 ...
 $ chas   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ nox    : num  0.538 0.469 0.469 0.458 0.458 0.458 0.524 0.524 0.524 0.524 ...
 $ rm     : num  6.58 6.42 7.18 7 7.15 ...
 $ age    : num  65.2 78.9 61.1 45.8 54.2 58.7 66.6 96.1 100 85.9 ...
 $ dis    : num  4.09 4.97 4.97 6.06 6.06 ...
 $ rad    : int  1 2 2 3 3 3 5 5 5 5 ...
 $ tax    : num  296 242 242 222 222 222 311 311 311 311 ...
 $ ptratio: num  15.3 17.8 17.8 18.7 18.7 18.7 15.2 15.2 15.2 15.2 ...
 $ black  : num  397 397 393 395 397 ...
 $ lstat  : num  4.98 9.14 4.03 2.94 5.33 ...
 $ medv   : num  24 21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 ...


In [34]:
mu <- mean(Boston$medv)
round (mu, 3)

[1] 22.533

### 9b) Provide an estimate of the standard error of ˆμ. Interpret this result.

In [32]:
mu_se <- sd(Boston$medv)/sqrt(length(Boston$medv))
round(mu_se, 3)

[1] 0.409

### 9c) Now estimate the standard error of ˆμ using the bootstrap

In [35]:
alpha.fn = function(data, index)
    return(mean(data[index]))
bootstrap <- boot(medv, alpha.fn, R=1000) 

alpha.fn
bootstrap

function(data, index)
    return(mean(data[index]))
<bytecode: 0x7fa8754b9df8>


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = alpha.fn, R = 1000)


Bootstrap Statistics :
    original      bias    std. error
t1* 22.53281 0.007650791   0.4106622

> When comparing the standard error from the population mean (0.409) and the one of bootstrap (0.402), one can say, that they are almost the same. 

### 9d) Provide a 95% con- fidence interval for the mean of medv & compare to t-test results.

In [36]:
lower_ci <- 22.53281-2*0.4023801
upper_ci <- 22.53281+2*0.4023801

lower_ci
upper_ci

t.test(Boston$medv)

[1] 21.72805

[1] 23.33757


	One Sample t-test

data:  Boston$medv
t = 55.111, df = 505, p-value < 2.2e-16
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
 21.72953 23.33608
sample estimates:
mean of x 
 22.53281 


> The mean of the bootstrap estimate was 22.533 with a CI of [21.728; 23.338], where as the results from the t-test show following: mean estimate of 22.533 with CI of [21.730; 23.336]. They are almost the same. 

 

### 9e) Provide an estimate, ˆμmed, for the median value of medv in the population.

In [37]:
medv.med = median(medv)
medv.med

[1] 21.2

### 9f) Estimate the standard error of the median using the bootstrap - Comment.

In [38]:
alpha.fn2 = function(data, index)
    return(median(data[index]))
bootstrap2 <- boot(medv, alpha.fn2, R=1000) 

alpha.fn2
bootstrap2

function(data, index)
    return(median(data[index]))
<bytecode: 0x7fa86cd770d0>


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = alpha.fn2, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*     21.2 -0.0386   0.3770241

> The standard error of mu computed with bootstrap is 0.388. The mean of 21.2 is the same as computed with median().


### 9g) Provide an estimate for the tenth percentile of medv in Boston suburbs.

In [39]:
medv.0.1 = quantile(medv, c(0.1))
medv.0.1

10% 
12.75

### 9h) Estimate the standard error of ˆμ0.1 using bootstrap.

In [40]:
alpha.fn3 = function(data, index)
    return(quantile(data[index], 0.1))
bootstrap3 <- boot(medv, alpha.fn3, R=1000) 

alpha.fn3
bootstrap3

function(data, index)
    return(quantile(data[index], 0.1))
<bytecode: 0x7fa8718cea20>


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = alpha.fn3, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*    12.75  0.0186   0.4925766

> The standard error of ˆμ0.1 computed with bootstrap is 0.493. The mean of 12.75 is the same as computed with quantile().